mlflow/utils/requirements_utils.py

```diff
@@ -426,7 +434,17 @@ def _infer_requirements(model_uri, flavor):
             _PYPI_PACKAGE_INDEX.date,
             unrecognized_packages,
         )
-    return sorted(map(_get_pinned_requirement, packages))
+
+    pinned_requirements = []
+    for package in packages:
+        # special casing for mlflow[gateway] to ensure proper formatting of the extra
+        if package == "mlflow[gateway]":
+            pinned_requirement = _get_pinned_requirement("mlflow", extras=["gateway"])
+        else:
+            pinned_requirement = _get_pinned_requirement(package)
+        pinned_requirements.append(pinned_requirement)
+
+    return sorted(pinned_requirements)
```

Comment:
if we want to keep the one liner i can make this something like

```
return sorted([
    _get_pinned_requirement(pkg) for pkg in packages 
    if pkg != "mlflow[gateway]" 
    else _get_pinned_requirements("mlflow", extras=["gateway"]
)])
```

Reply:
I like the current code. It looks more readable than one-liner.

Reply:
Another option is refactor `uninstalled` to accept a package requirement like `mlflow[gateway]`. This allows us to remove `if package == "mlflow[gateway]`.

Reply:
this makes sense, i went with this solution in the end!